jina/peapods/runtimes/gateway/prefetch.py

```diff
@@ -1,22 +1,57 @@
 import argparse
 import asyncio
-from abc import ABC
-from typing import AsyncGenerator
+from asyncio import Future
+from typing import AsyncGenerator, Dict
 
-from ....helper import typename
+from ....helper import typename, get_or_reuse_loop
 from ....logging.logger import JinaLogger
 from ....types.message import Message
 
-__all__ = ['PrefetchCaller', 'PrefetchMixin']
+__all__ = ['PrefetchCaller']
 
 if False:
     from ...zmq import AsyncZmqlet
 
 
-class PrefetchMixin(ABC):
-    """JinaRPCServicer """
+class PrefetchCaller:
+    """An async zmq request sender to be used in the Gateway"""
 
-    async def Call(self, request_iterator, *args) -> AsyncGenerator[None, Message]:
+    def __init__(self, args: argparse.Namespace, zmqlet: 'AsyncZmqlet'):
+        """
+        :param args: args from CLI
+        :param zmqlet: zeromq object
+        """
+        self.args = args
+        self.zmqlet = zmqlet
+        self.name = args.name or self.__class__.__name__
+
+        self.logger = JinaLogger(self.name, **vars(args))
+        self._message_buffer: Dict[str, Future[Message]] = dict()
+        self._is_running = True
```

Comment:
You want to wait on the event in the receive loop instead of checking the bool? Is there an advantage in doing this?

Reply:
Mainly cause `Events` should be used for synchronization. The receive loop spends most of its time awaiting on `recv_message`. When `close` sets `_is_running` to False, there's no assurance that receive can read that. The current logic works (primarily) because we `cancel` the `receive` task during close. 

I think we should use an Event for `is_running` or not use anything at all (use `task.cancel()` directly). Not sure of the best practice here.

Reply:
ok sure, Joan mentioned the same about `is_running` and I agree that its not really needed, the canceling should always work.